File: 2023/etc/dp/1006e-dp-230514.py
import sys
input = lambda: sys.stdin.readline().rstrip()

def dp():
	for i in range(2, N + 2):
		b0, b1 = (W >= A[0][i - 1] + A[0][i]), (W >= A[1][i - 1] + A[1][i])

		a0[i] = a1[i] = a[i - 1] + 1
		if b0:			a0[i] = min(a0[i], a1[i - 1] + 1)
		if b1:			a1[i] = min(a1[i], a0[i - 1] + 1)

		a[i] = min(a[i - 1] + (1 if W >= A[0][i] + A[1][i] else 2), a0[i] + 1, a1[i] + 1)
		if b0 and b1:	a[i] = min(a[i], a[i - 2] + 2)

T = int(input())
for _ in range(T):
	N, W = map(int, input().split())
	A = [ [ 0, 0 ] + list(map(int, input().split())) for _ in range(2) ]

	a, a0, a1 = [ [ 0 ] * (N + 2) for _ in range(3) ]

	dp()
	ans = a[-1]
	if N > 1:
		t, t0, t1 = W + 1, A[0][2], A[1][2]
		b0, b1 = (W >= A[0][2] + A[0][-1]), (W >= A[1][2] + A[1][-1])

		A[0][2] = A[1][2] = t; dp(); A[0][2], A[1][2] = t0, t1
		ans = min(ans, a[-2] + (0 if b0 and b1 else 2))

		A[0][2] = t; dp(); A[0][2] = t0
		ans = min(ans, a1[-1] + (0 if b0 else 1))

		A[1][2] = t; dp(); A[1][2] = t1
		ans = min(ans, a0[-1] + (0 if b1 else 1))
	print(ans)

File: boj/2023/tags/dp/1006E-dp-230514.c
#include	<stdio.h>
#define		MIN(x, y)	((x) > (y) ? (y) : (x))
#define		SZ			10003

int T, N, W;
int _A[2][SZ], *A[2] = { &_A[0][2], &_A[1][2] };
int a[SZ], a0[SZ], a1[SZ];
int ans;

void dp()
{
	int b0, b1;
	int i;

	for (i = 1; i <= N; i++) {
		b0 = (W >= A[0][i - 1] + A[0][i]), b1 = (W >= A[1][i - 1] + A[1][i]);

		a0[i] = a1[i] = a[i - 1] + 1;
		if (b0)			a0[i] = MIN(a0[i], a1[i - 1] + 1);
		if (b1)			a1[i] = MIN(a1[i], a0[i - 1] + 1);

		a[i] = MIN(a[i - 1] + (W >= A[0][i] + A[1][i] ? 1 : 2), MIN(a0[i] + 1, a1[i] + 1));
		if (b0 && b1)	a[i] = MIN(a[i], a[i - 2] + 2);
	}
}

int main()
{
	int t, t0, t1;
	int b0, b1;
	int i;

	scanf("%d", &T);
	while (T--) {
		scanf("%d %d", &N, &W);
		for (i = 1; i <= N; i++)	scanf("%d", &A[0][i]);
		for (i = 1; i <= N; i++)	scanf("%d", &A[1][i]);

		dp();
		ans = a[N];
		if (N > 1) {
			t = W + 1, t0 = A[0][1], t1 = A[1][1];
			b0 = (W >= A[0][1] + A[0][N]), b1 = (W >= A[1][1] + A[1][N]);

			A[0][1] = A[1][1] = t; dp(); A[0][1] = t0, A[1][1] = t1;
			ans = MIN(ans, a[N - 1] + (b0 && b1 ? 0 : 2));

			A[0][1] = t; dp(); A[0][1] = t0;
			ans = MIN(ans, a1[N] + (b0 ? 0 : 1));

			A[1][1] = t; dp(); A[1][1] = t1;
			ans = MIN(ans, a0[N] + (b1 ? 0 : 1));
		}
		printf("%d\n", ans);
	}
}